Fixed respawn anchor explode blocks underwater

b1zeyofficial opened the pull request:
As per title

### Related issues & PRs
 - Fixes https://github.com/pmmp/PocketMine-MP/issues/6716

### Behavioural changes
i think the servers behavior shouldnt change, but need to check

## Tests
https://www.youtube.com/watch?v=FK-LLQQMVEo

b1zeyofficial committed:
initial fix for this, it works

dktapps commented on src/block/RespawnAnchor.php:
```diff
@@ -110,14 +111,28 @@ private function explode(?Player $player) : void{
+			}
+		}
+		return false;
+	}
```
Why isn't this checking the down face?

b1zeyofficial replied:
I seem to have forgotten that the Java code checks for water under each horizontal block, since the existing code is similar to vanilla behavior.

P.S:
Apparently we should only check for horizontal blocks and not the Respawn Anchor itself, im confused
```java
	private static boolean isWaterThatWouldFlow(BlockPos blockPos, Level level) {
		FluidState fluidState = level.getFluidState(blockPos);
		if (!fluidState.is(FluidTags.WATER)) {
			return false;
		} else if (fluidState.isSource()) {
			return true;
		} else {
			float f = fluidState.getAmount();
			if (f < 2.0F) {
				return false;
			} else {
				FluidState fluidState2 = level.getFluidState(blockPos.below());
				return !fluidState2.is(FluidTags.WATER);
			}
		}
	}
	
	private void explode(BlockState blockState, Level level, BlockPos blockPos) {
		level.removeBlock(blockPos, false);
		boolean bl = Direction.Plane.HORIZONTAL.stream().map(blockPos::relative).anyMatch(blockPosx -> isWaterThatWouldFlow(blockPosx, level));
```

b1zeyofficial committed:
the logic for the "water detection" behavior has been changed

b1zeyofficial committed:
PHPStan

b1zeyofficial commented:
It doesnt really mimic vanilla behavior, for example, when water flows and a respawn anchor is placed in place, it doesnt explode blocks like it does in vanilla. Can we measure the fluidity of water using the Liquid->getDecay method?

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor-underwater

dktapps commented:
Yes, but decay is opposite to the fluid height. 0 is source, 7 is the shallowest.

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor-underwater

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor-underwater

b1zeyofficial committed:
Merge branch 'minor-next' into respawn-anchor-underwater

b1zeyofficial committed:
upgrade

b1zeyofficial commented:
everything remained the same, only the style changed and a couple of changes, it seems like I check that decay >= 1 but it still explodes after only 2 streams, maybe I miscalculated somewhere? 
I also removed `instanceof Liquid` because we dont want lava to be able to somehow muffle the anchor's explosion, do we?

dependabot[bot] committed:
Bump the github-actions group with 2 updates (#7019)

Bumps the github-actions group with 2 updates: [docker/build-push-action](https://github.com/docker/build-push-action) and [actions/upload-artifact](https://github.com/actions/upload-artifact).


Updates `docker/build-push-action` from 6.18.0 to 6.19.2
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/v6.18.0...v6.19.2)

Updates `actions/upload-artifact` from 6 to 7
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v6...v7)

---
updated-dependencies:
- dependency-name: docker/build-push-action
  dependency-version: 6.19.2
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: github-actions
- dependency-name: actions/upload-artifact
  dependency-version: '7'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Wraith0x10 committed:
Reduce glowstone count when interacting with a respawn anchor. (#7022)

dktapps committed:
LoginPacketHandler: make debug message less noisy

pmmp-admin-bot[bot] committed:
New translations from Crowdin (stable) (#7025)

Co-authored-by: pmmp-admin-bot[bot] <188621379+pmmp-admin-bot[bot]@users.noreply.github.com>

dktapps committed:
Update composer dependencies

dktapps committed:
CS

dktapps committed:
Prepare 5.41.1 release (#7026)

pmmp-admin-bot[bot] committed:
5.41.2 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/23322456108

dries-c committed:
Bedrock 1.26.10 (#7028)

pmmp-admin-bot[bot] committed:
5.42.1 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/23567046783

dktapps committed:
Reduce log spam from large packets

dktapps committed:
Update BedrockProtocol

dktapps committed:
Update PHPStan

dktapps committed:
Prepare 5.42.1 release (#7034)

pmmp-admin-bot[bot] committed:
5.42.2 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/23990887225

dependabot[bot] committed:
Bump the github-actions group across 1 directory with 7 updates (#7038)

Bumps the github-actions group with 7 updates in the / directory:

| Package | From | To |
| --- | --- | --- |
| [actions/create-github-app-token](https://github.com/actions/create-github-app-token) | `2` | `3` |
| [shivammathur/setup-php](https://github.com/shivammathur/setup-php) | `2.36.0` | `2.37.0` |
| [docker/setup-buildx-action](https://github.com/docker/setup-buildx-action) | `3` | `4` |
| [docker/build-push-action](https://github.com/docker/build-push-action) | `6.19.2` | `7.1.0` |
| [docker/login-action](https://github.com/docker/login-action) | `3` | `4` |
| [ncipollo/release-action](https://github.com/ncipollo/release-action) | `1.20.0` | `1.21.0` |
| [actions/github-script](https://github.com/actions/github-script) | `8` | `9` |



Updates `actions/create-github-app-token` from 2 to 3
- [Release notes](https://github.com/actions/create-github-app-token/releases)
- [Commits](https://github.com/actions/create-github-app-token/compare/v2...v3)

Updates `shivammathur/setup-php` from 2.36.0 to 2.37.0
- [Release notes](https://github.com/shivammathur/setup-php/releases)
- [Commits](https://github.com/shivammathur/setup-php/compare/2.36.0...2.37.0)

Updates `docker/setup-buildx-action` from 3 to 4
- [Release notes](https://github.com/docker/setup-buildx-action/releases)
- [Commits](https://github.com/docker/setup-buildx-action/compare/v3...v4)

Updates `docker/build-push-action` from 6.19.2 to 7.1.0
- [Release notes](https://github.com/docker/build-push-action/releases)
- [Commits](https://github.com/docker/build-push-action/compare/v6.19.2...v7.1.0)

Updates `docker/login-action` from 3 to 4
- [Release notes](https://github.com/docker/login-action/releases)
- [Commits](https://github.com/docker/login-action/compare/v3...v4)

Updates `ncipollo/release-action` from 1.20.0 to 1.21.0
- [Release notes](https://github.com/ncipollo/release-action/releases)
- [Commits](https://github.com/ncipollo/release-action/compare/v1.20.0...v1.21.0)

Updates `actions/github-script` from 8 to 9
- [Release notes](https://github.com/actions/github-script/releases)
- [Commits](https://github.com/actions/github-script/compare/v8...v9)

---
updated-dependencies:
- dependency-name: actions/create-github-app-token
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
- dependency-name: shivammathur/setup-php
  dependency-version: 2.37.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: github-actions
- dependency-name: docker/setup-buildx-action
  dependency-version: '4'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
- dependency-name: docker/build-push-action
  dependency-version: 7.1.0
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
- dependency-name: docker/login-action
  dependency-version: '4'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
- dependency-name: ncipollo/release-action
  dependency-version: 1.21.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: github-actions
- dependency-name: actions/github-script
  dependency-version: '9'
  dependency-type: direct:production
  update-type: version-update:semver-major
  dependency-group: github-actions
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dktapps committed:
Bedrock 1.26.20 (#7040)

pmmp-admin-bot[bot] committed:
5.43.1 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/25604740770

dktapps committed:
Fixed self-signed authentication for 1.26.20

dktapps committed:
Add missing file
Same shit different year

dktapps committed:
I hate this stupid CS check

dktapps committed:
5.43.1 (#7042)

pmmp-admin-bot[bot] committed:
5.43.2 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/25607562446

dependabot[bot] committed:
Bump phpstan/phpstan-strict-rules in the phpstan group (#7039)

Bumps the phpstan group with 1 update: [phpstan/phpstan-strict-rules](https://github.com/phpstan/phpstan-strict-rules).


Updates `phpstan/phpstan-strict-rules` from 2.0.10 to 2.0.11
- [Release notes](https://github.com/phpstan/phpstan-strict-rules/releases)
- [Commits](https://github.com/phpstan/phpstan-strict-rules/compare/2.0.10...2.0.11)

---
updated-dependencies:
- dependency-name: phpstan/phpstan-strict-rules
  dependency-version: 2.0.11
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: phpstan
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dktapps committed:
Player: fix several bed bugs

- player becoming invisible when leaving bed far from 0,0,0
- player staying stuck in sleep state when bed is destroyed

closes #7000

dktapps committed:
Fix FullContainerName handling in protocol
vanilla always seems to send this as null for slots, and 0+null for content.
I presume it's different if bundles are involved, but I didn't test that.

hopefully fixes #7043

dktapps committed:
Fixed spammy placement actions getting filtered
in newer versions the game sometimes sends all the same data with a different click face, with coordinates that make no sense. We need to make sure the correct ones don't get blocked by the packet spam filter.

closes #7014

dktapps committed:
5.43.2 release (#7046)

pmmp-admin-bot[bot] committed:
5.43.3 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/26727366211

dktapps committed:
no idea why this is suddenly broken but pls stop spamming my notifications

dependabot[bot] committed:
Bump shivammathur/setup-php from 2.37.0 to 2.37.1 in /.github/workflows (#7044)

Bumps [shivammathur/setup-php](https://github.com/shivammathur/setup-php) from 2.37.0 to 2.37.1.
- [Release notes](https://github.com/shivammathur/setup-php/releases)
- [Commits](https://github.com/shivammathur/setup-php/compare/2.37.0...2.37.1)

---
updated-dependencies:
- dependency-name: shivammathur/setup-php
  dependency-version: 2.37.1
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dktapps committed:
Bedrock 1.26.30 (#7050)

pmmp-admin-bot[bot] committed:
5.44.1 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/27981208174

dktapps committed:
InGamePacketHandler: guard against negative page numbers

dktapps committed:
InGamePacketHandler: guard against negative page numbers

dktapps committed:
5.44.1

pmmp-admin-bot[bot] committed:
5.44.2 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/27982146823

dktapps committed:
Merge branch 'stable' of github.com:pmmp/PocketMine-MP into stable

dktapps committed:
ResourcePackPacketHandler: fixed completion handler allowing multiple entries

dktapps committed:
5.44.2 release (#7053)

pmmp-admin-bot[bot] committed:
5.44.3 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/28131231060

dktapps committed:
a bunch of stuff php-cs-fixer wanted to mess with for no reason

xRookieFight committed:
fix: assign unique net IDs to furnace recipes

dktapps committed:
Update composer dependencies

dktapps committed:
Ready 5.44.3 (#7057)

pmmp-admin-bot[bot] committed:
5.44.4 is next

Commit created by: https://github.com/pmmp/RestrictedActions/actions/runs/28822675489

dktapps committed:
End of support announcement

b1zeyofficial committed:
Merge branch 'pmmp:stable' into respawn-anchor-underwater
